fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/EtagCacheRequestHandler.java

```diff
@@ -0,0 +1,141 @@
+     * Build a Cache request handler.
+     *
+     * @param next  The next handler in the chain
+     * @param dataCache  The cache instance
+     * @param querySigningService  The service to generate query signatures
+     * @param mapper  The mapper for all JSON processing
+     */
+    public EtagCacheRequestHandler(
+            @NotNull DataRequestHandler next,
+            @NotNull DataCache<?> dataCache,
+            @NotNull QuerySigningService<?> querySigningService,
+            @NotNull ObjectMapper mapper
+    ) {
+        super(mapper);
+        this.next = next;
+        this.dataCache = (TupleDataCache<String, String, String>) dataCache;
+        this.querySigningService = (QuerySigningService<Long>) querySigningService;
+    }
+
+    @Override
+    public boolean handleRequest(
+            RequestContext context,
+            DataApiRequest request,
+            DruidAggregationQuery<?> druidQuery,
+            ResponseProcessor response
+    ) {
+        ResponseProcessor nextResponse = response;
+
+        String cacheKey = null;
+        try {
+            if (context.isReadCache()) {
+                cacheKey = getKey(druidQuery);
+                final TupleDataCache.DataEntry<String, String , String> cacheEntry = dataCache.get(cacheKey);
+                CACHE_REQUESTS.mark(1);
+
+                String eTagInRequest = DruidJsonResponseContentKeys.REQUEST_ETAG.getName();
+                if (cacheEntry != null) { // Current query is in data cache
+                    // Insert "If-None-Match" header into RequestContext; the value is etag of the corresponding cache
+                    context.getHeaders().putSingle(
+                            eTagInRequest,
+                            mapper.readTree(cacheEntry.getValue()).get(DruidJsonResponseContentKeys.ETAG.getName())
+                                    .asText()
+                            );
+
+                    if (context.getNumberOfOutgoing().decrementAndGet() == 0) {
+                        RequestLog.record(new BardQueryInfo(druidQuery.getQueryType().toJson(), true));
+                        RequestLog.stopTiming(REQUEST_WORKFLOW_TIMER);
+                    }
+
+                    if (context.getNumberOfIncoming().decrementAndGet() == 0) {
+                        RequestLog.startTiming(RESPONSE_WORKFLOW_TIMER);
+                    }
+                } else { // Current query is not in data cache
+                    // Insert "If-None-Match" header into RequestContext; the value a random pre-defined string
+                    context.getHeaders().putSingle(
+                            eTagInRequest,
+                            DruidJsonResponseContentKeys.NON_EXISTING_ETAG_VALUE.getName()
+                    );
+
+                    CACHE_MISSES.mark(1);
+                }
+            }
+        } catch (Exception e) {
+            LOG.warn("Cache key cannot be built: ", e);
+        }
+
+        nextResponse = new EtagCacheResponseProcessor(
```

Comment:
So... it is possible that we don't put the etag header and still have the `EtagResponseProcessor` in the response chain when `isReadCache` is false?